### DISCUSSION TITLE
Customize category and tag pages + improving slugs

### DISCUSSION THREAD

**Discussion by JeremieLitzler** [vuepress-theme-hope/vuepress-theme-hope#3274] [Ideas]
Hi @Mister-Hope 

I hope you're doing well.

I'm not sure if I can do it, but, in French, categories or tags like `Santé` or `L'Archipelle` don't get evaluated properly.

The issue is that the characters get encoded and the [RFC3986 says no](https://www.rfc-editor.org/rfc/rfc3986#section-2), that is, URIs cannot contain unicode characters, only ASCII.

The problem is on the accentuated characters or same specials characters like the apostrophe.

Is that an enhancement that you can do? Or can you point me in the location of the code to update to handle that ([here?](https://github.com/vuepress-theme-hope/vuepress-theme-hope/tree/main/packages/blog2/src)) ?

I think that using the normalization method could work: 

```js
//for accents
const str = "Santé";
const cleaned = str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
//cleaned => "Sante"

//for special characters
const str2 = "L'archipelle";
const cleaned2 = str2.replace(/[.,\\\/#!$%\^&\*;:{}=\-_\'`~\(\)\"[\|]/g, "-");
//cleaned2 => "L-archipelle"
```

Also, in French, we say "Catégorie Santé", not "Santé Catégorie". Where can we update that in the page title? Maybe we could use a single method by printing "Catégorie: Santé" or "Category: Health" and it could be good enough for everyone?

Thanks very much.

**Comment by Mister-Hope** [comment]
I think that blog plugin should encode them well, that should be a bug.

Edited: Maybe there are some problems, I will check thay out later.

**Reply by JeremieLitzler** [reply]
Thanks @Mister-Hope 

I adde an extra request about page title. What's your position on that?

> Also, in French, we say "Catégorie Santé", not "Santé Catégorie". Where can we update that in the page title? Maybe we could use a single method by printing "Catégorie: Santé" or "Category: Health" and it could be good enough for everyone?

**Reply by Mister-Hope** [reply]
I am confusing about those words that I do not understand the meaning, can you point out exactly (maybe with screenshots) to describe what's happening and how would you like to improve it?

**Reply by Mister-Hope** [reply]
Oh are you talking about the tab title?
![image](https://github.com/vuepress-theme-hope/vuepress-theme-hope/assets/33315834/58de55f5-2373-4f42-9b01-00442414e191)

**Reply by Mister-Hope** [reply]
![image](https://github.com/vuepress-theme-hope/vuepress-theme-hope/assets/33315834/a2cc9cd0-7a5d-4c3e-b91b-8e0cb95635c9)

It's a bit complicated to do such things, as we must include extra locale options for these kinds of requests, and I don't think it's worse to do so.

If we are serious about these i18n, then we must provide these locale options:

- category (E.g.: `Category`)
- categories (E.g.: `Categories`)
- itemCategory (E.g.: `Category :item`)

With tag, star, article, timeline and other types, though we have a nicer default i18n now, it would be complicated for both developer and users who want to customize them.

**Reply by Mister-Hope** [reply]
So my basic point is that I prefer to provide less options (i.e.: only a single `category` locale option for category)

**Reply by JeremieLitzler** [reply]
@Mister-Hope 

> Oh are you talking about the tab title? ![image](https://user-images.githubusercontent.com/33315834/250806445-58de55f5-2373-4f42-9b01-00442414e191.png)

Yes, that's the one.

> So my basic point is that I prefer to provide less options (i.e.: only a single `category` locale option for category)

Well, in the case of `${name} ${themeData.locales[localePath].blogLocales.category}`, you output `Santé Catégorie` in French and `Health Category` in English.

Using `${themeData.locales[localePath].blogLocales.category} - ${name}`, you output `Catégorie - Santé` in French and `Category - Health` in English.

We could also have `${themeData.locales[localePath].blogLocales.category}: ${name}` or `${themeData.locales[localePath].blogLocales.category} / ${name}` or any delimiter character...

Same for the tags.

I think that it could be voted from the current users of the theme, what do you think?

**Reply by Mister-Hope** [reply]
> Using ${themeData.locales[localePath].blogLocales.category} - ${name}, you output Catégorie - Santé in French and Category - Health in English.

> We could also have ${themeData.locales[localePath].blogLocales.category}: ${name} or ${themeData.locales[localePath].blogLocales.category} / ${name} or any delimiter character...

Makes sense

**Comment by Mister-Hope** [comment]
https://github.com/vuepress-theme-hope/vuepress-theme-hope/commit/587eb6c0a4a1c9a1e6137d285a234c182d93ef16 fixes your problem

**Comment by JeremieLitzler** [comment]
Hi @Mister-Hope 

> [587eb6c](https://github.com/vuepress-theme-hope/vuepress-theme-hope/commit/587eb6c0a4a1c9a1e6137d285a234c182d93ef16) fixes your problem

I'm sorry, but this doesn't fix the issue I had in mind.

I was talking about replacing the UNICODE accentuated characters with ASCII characters (ex: `é` > `e ou `ô` > `o`) and also replacing any special characters to the dash character to make a pretty slug value (`'` > `-` or `?` > `-`).

I think the slug should only contain [a-zA-Z0-9] and the "-" character.

Is that a request that is doable / acceptable?

I'd understand it could become a complex matter.

Thanks for taking the time to read this.

**Reply by Mister-Hope** [reply]
the slugify function should be customizable in theme. Check that and open a pr.to.update it's default value.

**Comment by JeremieLitzler** [comment]
I see. I'll try that

**Comment by JeremieLitzler** [comment]
@Mister-Hope 

Are you sure it exposed to customization?

![image](https://github.com/vuepress-theme-hope/vuepress-theme-hope/assets/2410556/122b3645-7c4b-42a8-b28a-bb4fef906833)

I don't see it in beta.231.

**Reply by JeremieLitzler** [reply]
This is a screenshot from `theme.ts`

**Reply by Mister-Hope** [reply]
You can open a pr to support this in theme, and you can also improve the default slugify function

**Reply by Mister-Hope** [reply]
Busy in the passing days, have I support this yet?

**Reply by JeremieLitzler** [reply]
Hi @Mister-Hope 

I have no need of this anymore, having migrated to Hugo, which support out-of-the-box this feature.

Thanks.
